core/src/main/java/org/opensearch/sql/planner/physical/PhysicalPlan.java

```diff
@@ -43,6 +44,34 @@ public void add(Split split) {
 
   public ExecutionEngine.Schema schema() {
     throw new IllegalStateException(String.format("[BUG] schema can been only applied to "
-        + "ProjectOperator, instead of %s", toString()));
+        + "ProjectOperator, instead of %s", this.getClass().getSimpleName()));
+  }
+
+  /**
+   * Returns Total hits matched the search criteria. Note: query may return less if limited.
+   * {@see Settings#QUERY_SIZE_LIMIT}.
+   * Any plan which adds/removes rows to the response should overwrite it to provide valid values.
+   *
+   * @return Total hits matched the search criteria.
+   */
+  public long getTotalHits() {
+    return getChild().stream().mapToLong(PhysicalPlan::getTotalHits).max().orElse(0);
+  }
+
+  public String toCursor() {
+    throw new IllegalStateException(String.format("%s is not compatible with cursor feature",
+        this.getClass().getSimpleName()));
+  }
```

Comment:
What does these 2 new API mean for any physical operator? I guess you want an external visitor for these?

Reply:
`getTotalHits` -- total number of rows returned by an operator. Required to correctly return the total number of rows in the response to the query, like the legacy engine. Before this change, formatter returned query response size as total size of the query. For paginated responses this was returning fetch_size. 

`toCursor` allows an operator to be serialized and resumed later. This has been changed in https://github.com/opensearch-project/sql/pull/1498.

Reply:
I'm thinking should we do both in external visitor instead of adding the behavior to fundamental data structure (each physical plan node)

Reply:
`toCursor` is completely reworked and replaced by two other functions in #1498. It can't be moved to a visitor.
Having `getTotalHits` in a visitor requires access to private fields, but it is doable.

Reply:
thanks, it makes sense! I'm also trying to recall what's the principle for adding behavior to data structure or visitor in Visitor Design Pattern.